src/error.rs

```diff
@@ -40,9 +47,15 @@ pub enum Error {
     /// Input-output error. Apart from WouldBlock, these are generally errors with the
     /// underlying connection and you should probably consider them fatal.
     Io(io::Error),
-    #[cfg(feature = "tls")]
+    #[cfg(feature = "use-native-tls")]
     /// TLS error
     Tls(tls::Error),
+    #[cfg(feature = "use-rustls")]
```

Comment:
This is suboptimal. cargo features are supposed to be additive but enabling both here would cause compiler errors.

You can work around that by making the conditions a bit more complicated so that only one is actually used if both are enabled, but that would still cause problems if in the whole dependency chain one crate wants one feature and the other wants the other one.

The only proper solution to this is to reorganize things in a way that both features can be successfully enabled at once.

Reply:
That could indeed be a problem, didn't think about that. Most of the time you want either one of them but in bigger projects where you have multiple client to some 3rd party APIs that all use tungstenite, not all of them might give the option to select the TLS library.
How likely is that to happen though? Most libraries allow to pick or just transfer the decision to the user.

I saw that you use a fallback mechanism in `async-tungstenite` where you default to `native-tls` or `rustls` depending on what features are enabled.

Best would be to be explicit I guess but that would put an extra burden on the user for situations that may not happen too often. What I mean is we couldn't simple accept `wss://` URLs in `connect` anymore but would always require to pass extra information about the TLS implementation to use. That could be either a new parameter or have a separate `connect_tls` function maybe.
The disadvantage is that you would always have to pass the parameter like `connect_tls("wss://example.com", TlsConnector::Rustls);` even though you mostly only enable one TLS backend so you wouldn't really have to pass it and it only solves the rare case of having both backends enabled.

Maybe there could be a middleground where `connect` simply falls back to a pre-defined default, like always picking `native-tls` if both are enabled and add a new field to the config for `connect_with_config` that allows to be explicit.
That could be the most elegant version eventually, but can cause hidden side-effects if you rely on specific features of a TLS connector (for example `rustls` doesn't support doesn't support IP SANs yet) as the default may be different if both features happen to be enabled in user's crate and you can only circumvent it by always be explicit (and you may not even know about that detail).

That are my thoughts on this so far. What do you think @application-developer-DA ?

Reply:
>  How likely is that to happen though? Most libraries allow to pick or just transfer the decision to the user.

In a bigger application you might have two dependencies that as part of their job talk to some websocket API, and both could potentially select different TLS libraries for arbitrary reasons. It's not very likely but it's also very annoying if it happens :)

The main problem here is API surface that is changing in incompatible ways depending on the features selected.

Reply:
So I spend quite a few hours now shelling out a new variant of this changes, allowing both features to coexist and have a nice switch between no TLS at all and one or more TLS features enabled.

There are still several rough edges but it's very late for me and I will stop at this point for today. Would be great if you can take a look and tell me what you think @sdroege : https://github.com/snapview/tungstenite-rs/compare/master...dnaka91:rustls-multi-feature?expand=1

As you will see the changes are rather large to make it work even though I tried to keep changes at a minimum. The amount of extra feature flag checks is quite unfortunate as well.

Whenever you have time @application-developer-DA please let me know your opinions on this topic as well.

Reply:
@sdroege I updated this PR to allow both features at the same time now.

Now if both are enabled, then both errors will be enabled on the `Error` enum and some trait implementations are enabled as well. But for the actual TLS only `native-tls` will be used as auto-fallback. I would personally prefer `rustls` but the fact that `webpki` doesn't support IP SNIs yet makes it more likely to break clients if it would switch over.

The only pitfall currently is that someone might check for TLS errors from `rustls` but will never catch them as the crate would switch over to `native-tls`.

Maybe this could be merged as a first version of `rustls` support and eventually have the explicit variant in the future. What do you think?

Reply:
@dnaka91 the changes look good. Apart from the issues that @sdroege noticed I don't see any other issues with the PR. Do you plan to implement the suggestions that you mentioned to avoid issues in cases that @sdroege mentioned? I'm asking, so that I know if we can review it considering it to be some sort of a "final state" of the PR so to say :)

Reply:
Hey @application-developer-DA. So currently this PR should be complete, except I noticed some function docs mention the usage of native-tls that may need some update.

I adjusted the features so that both can be enabled at the same time and if they are then `native-tls` is used, but other things like the `TlsError` enum will have things from both crates so that no sudden compile errors happen.

My other idea was to make TLS usage fully explicit but that would require some bigger API changes. I thought to have `connect` for just plain connections without TLS and without any detection bases on the URL schema. Instead only accept `ws://` and no `wss://` then the `wss://` would only be allowed in a `connect_tls` function that accepts a `Tls` enum where we can pick which one to use. Alternatively have just two functions `connect_tls_native` and `connect_tls_rustls`.

I tried these changes on a separate branch and the differences are quite big (tried something here but this is not complete yet: https://github.com/snapview/tungstenite-rs/compare/master...dnaka91:rustls-multi-feature?expand=1).

Reply:
I believe this thread can be resolved now, as the features are now additive (with other comments dealing with some corner-cases). I don't believe the bigger changes for explicit TLS usage are needed now.